pkg/agent/utils.go

```diff
@@ -586,3 +586,25 @@ func addFeatureGateString(featureGates string, key string, value bool) string {
 	}
 	return strings.Join(pairs, ",")
 }
+
+// ParseCSE parses the raw VMSS CSE output
+func ParseCSE(message string) (datamodel.VMSSInstanceViewCSEStatus, error) {
+	var cseStatus datamodel.VMSSInstanceViewCSEStatus
+	var rerr error
+	start := strings.Index(message, "[stdout]") + len("[stdout]")
+	end := strings.Index(message, "[stderr]")
+	if end > start {
+		rawInstanceViewInfo := message[start:end]
+		err := json.Unmarshal([]byte(rawInstanceViewInfo), &cseStatus)
+		if err != nil || cseStatus.ExitCode == "" {
+			// Regex "vmssInstanceErrorCode=" is part of contract to parse the error, please inform clients who are relying on it before change the regex.
+			rerr = fmt.Errorf("vmssCSE has invalid message=%s, %s=%s", message, VMSSInstanceErrorCode, InvalidCSEMessage)
+		} else {
+			cseStatus.ExitCode = strings.Trim(cseStatus.ExitCode, "\"")
+		}
+	} else {
+		// Regex "vmssInstanceErrorCode=" is part of contract to parse the error, please inform clients who are relying on it before change the regex.
```

Comment:
no we should not bear the contract in an error message. We can introduce a new error type and return different errors for no message vs unmarshal errors. Client can use that to tell what is wrong.

Reply:
For example, when an error "CSE has invalid message=xx, InstanceErrorCode=CSEMessageUnmarshalError" is returned, this error will be finally parsed in core/armerror/deploymenterrorparser.go, we will check the regex pattern "InstanceErrorCode=" to parse the error code.